File: src/main/java/org/radarcns/gateway/util/ServletInputStreamWrapper.java
package org.radarcns.gateway.util;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import javax.servlet.ServletInputStream;

/**
 * ServletInputStream wrapper for an InputStream.
 */
public class ServletInputStreamWrapper extends ServletInputStream {
    private final InputStream stream;

    public ServletInputStreamWrapper(InputStream stream) throws IOException {
        this.stream = stream;
    }

    @Override
    public int read() throws IOException {
        return stream.read();
    }

    @Override
    public int read(byte[] buf, int off, int len) throws IOException {
        return stream.read(buf, off, len);
    }

    @Override
    public long skip(long off) throws IOException {
        return stream.skip(off);
    }

    @Override
    public void mark(int readLimit) {
        stream.mark(readLimit);
    }

    @Override
    public boolean markSupported() {
        return stream.markSupported();
    }

    @Override
    public void reset() throws IOException {
        stream.reset();
    }

    @Override
    public void close() throws IOException {
        stream.close();
    }

    /** read all bytes from a stream. Will not return on infinite streams. */
    public static byte[] readFully(InputStream stream) throws IOException {
        try (ByteArrayOutputStream dataStream = new ByteArrayOutputStream()) {
            byte[] buffer = new byte[4096];
            int numRead;
            while ((numRead = stream.read(buffer, 0, buffer.length)) != -1) {
                dataStream.write(buffer, 0, numRead);
            }
            return dataStream.toByteArray();
        }
    }
}

File: src/main/kotlin/org/radarcns/gateway/util/ServletInputStreamWrapper.kt
package org.radarcns.gateway.util

import java.io.IOException
import java.io.InputStream
import javax.servlet.ReadListener
import javax.servlet.ServletInputStream

/**
 * ServletInputStream wrapper for an InputStream.
 */
class ServletInputStreamWrapper @Throws(IOException::class)
constructor(private val stream: InputStream, private val baseStream: ServletInputStream) : ServletInputStream() {
    @Throws(IOException::class)
    override fun read() = stream.read()

    @Throws(IOException::class)
    override fun read(buf: ByteArray, off: Int, len: Int) = stream.read(buf, off, len)

    @Throws(IOException::class)
    override fun skip(off: Long) = stream.skip(off)

    override fun mark(readLimit: Int) = stream.mark(readLimit)

    override fun markSupported() = stream.markSupported()

    @Throws(IOException::class)
    override fun reset() = stream.reset()

    @Throws(IOException::class)
    override fun close() = stream.close()

    @Throws(IOException::class)
    override fun available() = stream.available()

    override fun isReady() = baseStream.isReady

    override fun isFinished() = baseStream.isFinished

    override fun setReadListener(readListener: ReadListener?) = baseStream.setReadListener(readListener)
}
